### PR TITLE
UCT/CUDA/GDR_COPY:   gdr copy uct

### PR DESCRIPTION
**Pull request by bureddy**
- copy from system memory to cuda memory using gdr copy utility
        - gdr copy (https://github.com/NVIDIA/gdrcopy)

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4797/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@bureddy has to be up to 500 lines, pls also fix commit message title

**Comment by bureddy** [comment]
@yosefe updated PR only with skeleton. will push remaining uct changes after this is in

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4849/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4868/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4972/ for details (Mellanox internal link).

File: src/uct/cuda/gdr_copy/gdr_copy_ep.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+
+#ifndef UCT_SYSV_EP_H
+#define UCT_SYSV_EP_H
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_ep.h]: UCT_GDR_COPY_EP_H

File: src/uct/cuda/gdr_copy/gdr_copy_ep.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+
+typedef struct uct_gdr_copy_ep_addr {
+    int                ep_id;
+} uct_gdr_copy_ep_addr_t;
+
+typedef struct uct_gdr_copy_ep {
+    uct_base_ep_t      super;
+    struct uct_gdr_copy_ep *next;
+} uct_gdr_copy_ep_t;
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_ep.h]: style: align struct fields

File: src/uct/cuda/gdr_copy/gdr_copy_iface.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCT_CUDA_IFACE_H
+#define UCT_CUDA_IFACE_H
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_iface.h]: UCT_CUDA_GDR_COPY_IFACE_H

File: src/uct/cuda/gdr_copy/gdr_copy_iface.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+static UCS_CLASS_INIT_FUNC(uct_gdr_copy_iface_t, uct_md_h md, uct_worker_h worker,
+                           const uct_iface_params_t *params,
+                           const uct_iface_config_t *tl_config)
+{
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_iface_t, &uct_gdr_copy_iface_ops, md, worker,
+                              params, tl_config UCS_STATS_ARG(params->stats_root)
+                              UCS_STATS_ARG(UCT_GDR_COPY_TL_NAME));
+
+    if (strcmp(params->mode.device.dev_name, UCT_CUDA_DEV_NAME) != 0) {
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_iface.c]: strncmp

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5029/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@alex-mikheev Please review,

File: src/uct/cuda/gdr_copy/gdr_copy_ep.c [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+#include <ucs/debug/memtrack.h>
+#include <ucs/type/class.h>
+
+
+static UCS_CLASS_INIT_FUNC(uct_gdr_copy_ep_t, uct_iface_t *tl_iface,
+                           const uct_device_addr_t *dev_addr,
+                           const uct_iface_addr_t *iface_addr)
+{
+    uct_gdr_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_gdr_copy_iface_t);
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super)
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_ep.c]: style: add extra line

File: src/uct/cuda/gdr_copy/gdr_copy_iface.c [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+/* Forward declaration for the delete function */
+static void UCS_CLASS_DELETE_FUNC_NAME(uct_gdr_copy_iface_t)(uct_iface_t*);
+
+
+static ucs_status_t uct_gdr_copy_iface_get_address(uct_iface_h tl_iface,
+                                                   uct_iface_addr_t *iface_addr)
+{
+    int *gdr_copy_addr = (int*)iface_addr;
+    *gdr_copy_addr = 0;
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_iface.c]: style: add extra line

File: src/uct/cuda/gdr_copy/gdr_copy_md.c [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+{
+    return UCS_OK;
+}
+
+static ucs_status_t uct_gdr_copy_mem_reg(uct_md_h md, void *address, size_t length,
+                                         unsigned flags, uct_mem_h *memh_p)
+{
+    ucs_status_t rc;
+    uct_mem_h * mem_hndl = NULL;
+    mem_hndl = ucs_malloc(sizeof(void *), "gdr_copy handle for test passing");
```
Comment (alex-mikheev) [inline src/uct/cuda/gdr_copy/gdr_copy_md.c]: style: add extra line between var decls and code

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5123/ for details (Mellanox internal link).

**Pull request merged**